### PR TITLE
fix(megatron): QK-Norm on MLA decoupled-RoPE to stop hybrid divergence

### PR DESCRIPTION
**Pull request by vanshbhatia-amd**
## What

Adds a **parameter-free QK-Norm** (unit-RMS `RMSNorm`) on the **decoupled-RoPE** q/k of Multi-Latent Attention, applied before `apply_rotary_pos_emb`.

## Why

In MLA the q/k are split into a **content** part (`qk_head_dim`) and a **decoupled-RoPE** part (`qk_pos_emb_head_dim`). The content part is normalized (`q_layernorm`/`kv_layernorm` on the compressed latents), but the **rotary part is never normalized** — it is projected, split off, and fed straight into `apply_rotary_pos_emb`.

In a KDA/MLA hybrid this is a divergence mechanism: the rotary component's magnitude grows with training until it dominates the per-head attention logits and blows up the deep residual stream. `qk_clip` (MuonClip) does **not** catch it — `qk_clip` only rescales the content projections, so the unnormalized rotary path escapes it entirely and the logits climb past any threshold.

Observed in a 3B KDA-RoPE hybrid: with `qk_clip` active and its threshold holding the *clipped* logit, training still diverged; per-layer instrumentation showed the deepest residual layers exploding together while the raw attention logits shot well past the clip threshold — i.e. the runaway came through the rotary path.

## Fix

Normalize `q_pos_emb` and `k_pos_emb` to **unit RMS** over `qk_pos_emb_head_dim` before RoPE. Since RoPE is norm-preserving, this bounds the rotary logit contribution to `<= qk_pos_emb_head_dim` regardless of projection-weight growth.

- **Parameter-free** (no learnable gain): adds no parameters, so the `state_dict` and distributed-optimizer shard layout are unchanged and **existing checkpoints resume unchanged**.
- **Targeted**: only the rotary (pos) component is normalized; the content path is untouched.

## Implementation

Source-string patch on `MLASelfAttention.get_query_key_value_tensors` (the non-fused path), consistent with the existing `patches/` convention. `PrimusMLASelfAttention` inherits the method unchanged, so patching the base class covers the hybrid.

- **Enable**: `mla_rope_qk_norm` config flag, or `PRIMUS_MLA_ROPE_QK_NORM=1` (no config-schema change required).
- **Scope**: non-fused RoPE path (`apply_rope_fusion: false`, `cache_mla_latents: false`). If `apply_rope_fusion=True` the patch logs a warning and no-ops (the fused kernel applies rotary internally).
- **Default off** → no behavior change for existing configs.

## Testing

Enabled on the 3B KDA-RoPE hybrid: the max attention logit dropped from ~5,600 to ~65, and training passed cleanly through the iteration where it previously diverged (grad flat ~0.1, zero grad-norm spikes), with loss at/below the pre-divergence baseline.

### CODE REVIEW

File: primus/backends/megatron/patches/mla_rope_qk_norm_patches.py [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+                k_pos_emb = k_pos_emb * torch.rsqrt(
+                    k_pos_emb.float().pow(2).mean(-1, keepdim=True) + _qkn_eps
+                ).to(k_pos_emb.dtype)
+
+                # q_pos_emb: [num_tokens, n, qk_pos_emb_head_dim]
+                q_pos_emb = apply_rotary_pos_emb("""
+
+
+def _install_mla_rope_qk_norm_patch() -> None:
+    from megatron.core.transformer.multi_latent_attention import MLASelfAttention
```
[the remark this answers is not in the document]
Comment (vanshbhatia-amd) [reply primus/backends/megatron/patches/mla_rope_qk_norm_patches.py:105]: Fixed — `_install` now resolves the class that defines `get_query_key_value_tensors` via the MRO, so the patch lands on the upstream base even if the module attribute was rebound to `PrimusMLASelfAttention`.

File: primus/backends/megatron/patches/mla_rope_qk_norm_patches.py [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+
+    if is_patched(MLASelfAttention, _PATCH_KEY):
+        log_rank_0(f"[Patch:{_PATCH_KEY}] already applied; skipping.")
+        return
+
+    patch_method_source(MLASelfAttention, "get_query_key_value_tensors", _ORI, _NEW)
```
[the remark this answers is not in the document]
Comment (vanshbhatia-amd) [reply primus/backends/megatron/patches/mla_rope_qk_norm_patches.py]: Added `tests/unit_tests/backends/megatron/test_mla_rope_qk_norm_patches.py` covering the upstream anchor, the rewrite + idempotency, and the enable condition (flag/env var + fused-RoPE guard).
